### ISSUE TITLE
Chaining operators for 'a option and 'a result

### ISSUE THREAD

**Issue by herry13** [ocaml-batteries-team/batteries-included#604]
Hi All, thanks for this great software. Question: is there any operator for chaining functions which return 'a option / 'a result? If not, I want to propose my implementations to be added into batteries which I found very useful and simple to be understood. The signatures of the functions are:

`val (|?>) : 'a option -> ('a -> 'a option) -> 'a option`

`val fold_opt : ('a -> 'b -> 'b option) -> 'a list -> 'b -> 'b option`

`val (||>) : ('a, exn) result -> ('a -> ('a, exn) result) -> ('a, exn) result`

`val fold_res : ('a -> 'b -> ('b, exn) result) -> 'a list -> 'b -> ('b, exn) result`

**Comment by hcarty** [comment]
For `(|?>)` we already have `BatOption.bind`.  For `(||>)` we have `BatResult.Monad.bind`.  For consistency they should both be spelled `(>>=)` which `result` already has as `BatResult.Infix.(>>=)`.  Adding `BatOption.Infix.(>>=)` seems like a reasonable thing to do.  `BatResult.Infix.(|?)` may make sense too.

I don't have an opinion one way or the other on the `fold` functions, other than that they should probably live in `BatList`.

**Comment by herry13** [comment]
@hcarty thanks for the pointers. I prefer to use different infix symbols so that a module can deal with `option` and `result`.

I agree for `fold` functions, they should be in `BatList`.

**Comment by hcarty** [comment]
For what it's worth, local module opens give you some flexibility with overlapping operator names.  That said, depending on the structure of your code they can make things pretty noisy.

**Comment by herry13** [comment]
@hcarty I understand that name consistency is important. However, my current project has functions that have to deal with both `option` and `result`. So overlapping operator names won't work in my case.

**Comment by gasche** [comment]
The right way to do those folds is to implement a folding function over any monad. @c-cube has implemented monad-generic `fold` (and `map` and `traverse`) for many sequence libraries (including lists) in the pull request #501 , but the code is sitting outside unmerged because @c-cube had no time to polish the code to meet the merge requirements (what it lacks are just unit tests). If anyone here is willing to implement the necessary unit tests, that would be a warmly appreciated contribution.

**Comment by herry13** [comment]
@gasche thanks for the comment. I tried @c-cube codes for `option` and it works:

```ocaml
module TOption = Traverse(BatOption.Monad)
let fold_opt = TOption.foldM
```

However, when I tried a similar way to `result`, it didn't work:

```ocaml
module TResult = Traverse(BatResult.Monad)
let fold_res = TResult.foldM
```

The error for the above code is:

```
Error: Signature mismatch:
       Modules do not match:
         sig
           val bind :
             ('a, 'b) BatResult.t ->
             ('a -> ('c, 'b) BatResult.t) -> ('c, 'b) BatResult.t
           val ( >>= ) :
             ('a, 'b) BatResult.t ->
             ('a -> ('c, 'b) BatResult.t) -> ('c, 'b) BatResult.t
           val return : 'a -> ('a, 'b) BatResult.t
         end
       is not included in
         BatInterfaces.Monad
       The type `m' is required but not provided
```

My knowledge on @c-cube codes are very limited. Do you know how to solve it?

**Comment by gasche** [comment]
The error is not caused by @c-cube patch but by a mistake in the `BatResult` code that I will fix as soon as possible.

**Comment by herry13** [comment]
@gasche could you please give me (or implement) your fix so that I can try it? I might have time to help you creating a "small" unit test for @c-cube patch by extracting some cases from my project.

**Comment by herry13** [comment]
As suggested by @hcarty, I've implemented BatOption.{Monad,Infix}.(>>=).

**Comment by gasche** [comment]
I thought the lack of a `BatInterfaces.Monad` instance in result was just an ommission, but in fact it's a bit more subtle than that -- the problem is that we have to "fix" the second parameter of `('a, 'b) Result.t` type to get a single-type-parameter monad.

I have implemented in a local branch [result-WithMonad](https://github.com/gasche/batteries-included/compare/result-WithMonad) a solution using a functor in `BatResult`. I'm worried that this approach is a bit too complex for users -- maybe we should just define `Monad2`, `Monad3` etc. for the simple cases.

**Comment by herry13** [comment]
For me, defining `Monad2`, `Monad3`, etc will be a better solution -- simple is better.

**Comment by gasche** [comment]
I may do that, and provide generic functors to convert from `MonadN` to `Monad1` using the technique used above.

**Comment by herry13** [comment]
Agree, it's a good solution. Any suggestion when I can get the implementation?

**Comment by herry13** [comment]
@gasche any news about the implementation of `Monad2`, `Monad3`, and `MonadN`?

**Comment by gasche** [comment]
Nothing yet, sorry. Thanks for pinging me: I need to find time to work on this, and the pings help. Do not hesitate to do it again.

**Comment by gasche** [comment]
I did another attack at the problem in my [result-WithMonad](https://github.com/gasche/batteries-included/tree/result-WithMonad) branch, with both `Monad1`, `Monad2` etc. and a (generic) functor converting to the usual Monad instance. Here is [the commit](https://github.com/gasche/batteries-included/commit/bca5ec56e5a1a2d7950087a55ee591c1c7f59450).

The code example are as follows:

```

(* monomorphic case: instantiate the functor with a fixed type *)
let sequence (seq : ('a, exn) Result.t Enum.t) : ('a Enum.t, exn) Result.t =
  let module M = Monad.Make1(Result.Monad)(struct type t = exn end) in
  let module RE = Enum.WithMonad(M) in
  RE.sequence seq

(* polymorphic case: instantiate the functor with a type parameter *)
let sequence (type b) (seq : ('a, b) Result.t Enum.t) : ('a Enum.t, b) Result.t =
  let module M = Monad.Make1(Result.Monad)(struct type t = b end) in
  let module RE = Enum.WithMonad(M) in
  RE.sequence seq
```

There is still a bit too much boilerplate to my taste, but the solution scales to other types than `BatResult`. I'm not sure how much more we could expect to hide the interaction between functors and parametrized datatypes to the user -- there may be no much-simpler solution in today's OCaml.
